### PR TITLE
Offline modules

### PR DESCRIPTION
**Pull request by szabinst**
### Offline modules:
When modules (or grades) is selected, every other tab will be selected except grades (or modules) since we most likely will need those to display module items properly. 

### Fixes
Fixed an issue where course sync selection got overridden overtime. 
Fixed an issue where redownloading already persisted files caused the progress banner to show incorrect numbers. 

---

https://github.com/instructure/canvas-ios/assets/110813321/1c3e4997-fa9c-4a66-810b-f074f4064b7b

---

Offline modules

refs: MBL-16888
affects: Student
release note: none
ticket: See ticket

## Checklist

- [x] Tested on phone
- [ ] Tested on tablet

### CODE REVIEW

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request will not generate a release note.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>One or more files are below the minimum test coverage 50%</td>
    </tr>
  </tbody>
</table>


#### Affected Apps: Student

[MBL-16888](https://instructure.atlassian.net/browse/MBL-16888)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Canvas iOS** | 92.39% | 92.44% | -0.05%
Core/Core/CourseSync/CourseSyncDownloader/Model/CourseSyncModulesInteractor.swift | 34.15% | -- | --
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d4a1b8c9a4e0878dedbfadaf55abaf0a12ff4e6e
</p>

File: Core/Core/CourseSync/Common/Models/CourseSyncListerInteractor.swift [hunk 1]
```diff
@@ -128,22 +128,6 @@ class CourseSyncListInteractorLive: CourseSyncListInteractor {
...
-            sessionDefaults.offlineSyncSelections = oldSelections + newSelections
-        case .all:
-            // If all courses are visible then it's safe to overwrite all course selections
-            sessionDefaults.offlineSyncSelections = CourseSyncItemSelection.make(from: entries)
-        case .synced:
-            break
-        case .none:
-            break
-        }
-
```
Comment (vargaat) [inline Core/Core/CourseSync/Common/Models/CourseSyncListerInteractor.swift:146]: How is this not required any more?

File: Core/Core/CourseSync/Common/Models/Entities/CourseSyncEntry.swift [hunk 2]
```diff
@@ -182,6 +182,14 @@ public struct CourseSyncEntry: Equatable {
...
         }
 
+        if selectionState == .selected {
+            if tabs[id: id]?.type == .modules || tabs[id: id]?.type == .grades {
+                for (index, tab) in tabs.enumerated() where tab.type != .modules && tab.type != .grades {
+                    tabs[index].selectionState = .selected
+                }
+            }
+        }
+
```
Comment (vargaat) [inline Core/Core/CourseSync/Common/Models/Entities/CourseSyncEntry.swift:192]: Could we add a logic that de-selects grades and modules if any of the dependent items are de-selected? Now if I select modules and after de-select manually every other item except modules then it won't work as expected.

**Review by vargaat — CHANGES_REQUESTED** [review]

File: Core/Core/CourseSync/Common/Models/CourseSyncListerInteractor.swift (hunk 1 above)
Comment (szabinst) [reply Core/Core/CourseSync/Common/Models/CourseSyncListerInteractor.swift:146]: This logic was copied and inserted here by mistake. The purpose of this interactor is to return a list of entries for sync selector/progress/scheduled download screens. The selection persistance logic is handled on the selector screen.

File: Core/Core/CourseSync/Common/Models/Entities/CourseSyncEntry.swift (hunk 2 above)
Comment (szabinst) [reply Core/Core/CourseSync/Common/Models/Entities/CourseSyncEntry.swift:192]: Good idea, added.

**Review by vargaat — APPROVED** [review]
QA+1

**Review by balintbartok — APPROVED** [review]
QA + 1

**Pull request merged**